core/lib/server/pub-op.db.test.ts

```diff
@@ -0,0 +1,1329 @@
+							values: {
+								Title: "A pub related to another Pub",
+							},
+						},
+					},
+				],
+			},
+		},
+		{
+			stage: "Stage 1",
+			pubType: "Minimal Pub",
+			values: {
+				Title: "Minimal pub",
+			},
+		},
+	],
+});
+
+let seededCommunity: CommunitySeedOutput<typeof seed>;
+
+beforeAll(async () => {
+	const { seedCommunity } = await import("~/prisma/seed/seedCommunity");
+	seededCommunity = await seedCommunity(seed);
+});
+
+describe("PubOp", () => {
+	it("should create a new pub", async () => {
+		const id = crypto.randomUUID() as PubsId;
+		const pubOp = PubOp.upsert(id, {
+			communityId: seededCommunity.community.id,
+			pubTypeId: seededCommunity.pubTypes["Basic Pub"].id,
+			lastModifiedBy: createLastModifiedBy("system"),
+		});
+
+		const pub = await pubOp.execute();
+		await expect(pub.id).toExist();
+	});
+
+	it("should not fail when upserting existing pub", async () => {
+		const id = crypto.randomUUID() as PubsId;
+		const pubOp = PubOp.upsert(id, {
+			communityId: seededCommunity.community.id,
+			pubTypeId: seededCommunity.pubTypes["Basic Pub"].id,
+			lastModifiedBy: createLastModifiedBy("system"),
+		});
+
+		const pub = await pubOp.execute();
+		await expect(pub.id).toExist();
+
+		const pub2 = await PubOp.upsert(id, {
+			communityId: seededCommunity.community.id,
+			pubTypeId: seededCommunity.pubTypes["Basic Pub"].id,
+			lastModifiedBy: createLastModifiedBy("system"),
+		}).execute();
+
+		await expect(pub2.id).toExist();
+	});
+
+	it("should create a new pub and set values", async () => {
+		const id = crypto.randomUUID() as PubsId;
+		const pubOp = PubOp.upsert(id, {
+			communityId: seededCommunity.community.id,
+			pubTypeId: seededCommunity.pubTypes["Basic Pub"].id,
+			lastModifiedBy: createLastModifiedBy("system"),
+		})
+			.set(seededCommunity.pubFields["Title"].slug, "Some title")
+			.set({
+				[seededCommunity.pubFields["Description"].slug]: "Some description",
+			});
+
+		const pub = await pubOp.execute();
+		await expect(pub.id).toExist();
+
+		expect(pub).toHaveValues([
+			{
+				fieldSlug: seededCommunity.pubFields["Description"].slug,
+				value: "Some description",
+			},
+			{
+				fieldSlug: seededCommunity.pubFields["Title"].slug,
+				value: "Some title",
+			},
+		]);
+	});
+
+	it("should be able to relate existing pubs", async () => {
+		const pubOp = PubOp.upsert(crypto.randomUUID() as PubsId, {
+			communityId: seededCommunity.community.id,
+			pubTypeId: seededCommunity.pubTypes["Basic Pub"].id,
+			lastModifiedBy: createLastModifiedBy("system"),
+		});
+
+		const pub = await pubOp.execute();
+
+		await expect(pub.id).toExist();
+
+		const pub2 = await PubOp.upsert(crypto.randomUUID() as PubsId, {
+			communityId: seededCommunity.community.id,
+			pubTypeId: seededCommunity.pubTypes["Basic Pub"].id,
+			lastModifiedBy: createLastModifiedBy("system"),
+		})
+			.relate(seededCommunity.pubFields["Some relation"].slug, "test relations value", pub.id)
+			.execute();
```

Comment:
this is quite nice! just so I understand, if pub2 were already created, would we also have to call `.upsert(...).relate(...)`? i.e. there's nothing like `PubOp.relate`

Reply:
Yeah correct!
You'd more likely do `PubOp.update(pub2Id).relate()` if you did not want to override the existing relations, but yes.

Reply:
got it, thanks for the context, makes sense!